gramlib/gramext.ml

```diff
@@ -469,55 +469,65 @@ and decr_keyw_use_in_tree gram =
       decr_keyw_use_in_tree gram n.son;
       decr_keyw_use_in_tree gram n.brother
 
-let rec delete_rule_in_suffix entry symbols =
-  function
+let rec delete_rule_in_suffix entry symbols = function
     lev :: levs ->
-      begin match delete_rule_in_tree entry symbols lev.lsuffix with
+    begin
+      match delete_rule_in_tree entry symbols lev.lsuffix with
         Some (dsl, t) ->
-          begin match dsl with
+        begin match dsl with
             Some dsl -> List.iter (decr_keyw_use entry.egram) dsl
           | None -> ()
-          end;
-          begin match t with
-            DeadEnd when lev.lprefix == DeadEnd -> levs
+        end;
+        begin match t with
+          (* We keep the level as requested by @herbelin *)
+          | DeadEnd when lev.lprefix == DeadEnd ->
+            let lev = { lev with assoc = NonA; lsuffix = t } in
+            lev :: levs
           | _ ->
-              let lev =
-                {assoc = lev.assoc; lname = lev.lname; lsuffix = t;
-                 lprefix = lev.lprefix}
-              in
-              lev :: levs
-          end
+            let lev =
```

Comment:
I suspect that you should remove this whole `DeadEnd` case. The bug in the test-suite is probably because `NonA` is used instead of `lev.assoc` (as in the non-`DeadEnd` case).

Idem in `delete_rule_in_prefix`.